### PR TITLE
Bring back automaton minimization

### PR DESCRIPTION
**Pull request by n1v0lg**
The security codebase relies heavily on automata and caching these. The Lucene 10 upgrade removed automaton minimization which can result in a memory usage increase of >5x, esp. for roles with many application privileges. 

This PR brings back Automaton minimization to avoid the explosion in roles cache size.

Relates: ES-10451

### CODE REVIEW

File: server/src/main/java/module-info.java
Comment (n1v0lg) [inline server/src/main/java/module-info.java:483]: Including this in `server` so that we can re-use minimization in the apm module as well: https://github.com/elastic/elasticsearch/pull/119344

File: server/src/main/java/org/elasticsearch/lucene/util/automaton/MinimizationOperations.java
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/lucene/util/automaton/MinimizationOperations.java:25]: Lets add a bit of background here on why we brought this back? 
I.e. "Lucene removed the minimization logic but Elasticsearch still requires it. Elasticsearch makes heavy use of automata for checking permissions in the security codebase which consumes a prohibitive amount of heap for automata without minimization."

**Review by original-brownbear — APPROVED** [review]
LGTM assuming CI is happy and maybe add a bit more comment to the new class before merging :)

Thanks!

File: server/src/main/java/org/elasticsearch/lucene/util/automaton/MinimizationOperations.java
Comment (n1v0lg) [reply server/src/main/java/org/elasticsearch/lucene/util/automaton/MinimizationOperations.java:29]: Oh ya, good call. Pushed https://github.com/elastic/elasticsearch/pull/119309/commits/eb8ef47f6e53f3551fe954fa09ddb774ef4bbe2a

**Pull request merged**
